### DISCUSSION TITLE
아이템 35. ordinal 메서드 대신 인스턴스 필드를 사용하라

### DISCUSSION THREAD

**Discussion by JoisFe** [Study-2-Effective-Java/Effective-Java#94] [3. 과제] [answered]
# 아이템 35. ordinal 메서드 대신 인스턴스 필드를 사용하라

### 열거 타입 상수의 특징
- 대부분 열거 타입 상수는 자연스럽게 하나의 정숫값에 대응
- 모든 열거 타입은 해당 상수가 그 열거 타입에서 몇 번째 위치인지를 반환하는 ordinal 메서드를 제공
- [참고 java11 doc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Enum.html) 
<img width="1170" alt="image" src="https://user-images.githubusercontent.com/90208100/215309291-c3f2cdef-6d49-4a00-86f7-e540ab6ffe9d.png">

### ordinal 메서드 사용의 유혹
- 열거 타입 상수와 연결된 정숫값이 필요한 경우 ordinal 메서드를 이용하고 싶어짐

### ordinal 메서드 사용의 문제점

``` java
public enum Ensemble {
    SOLO, DUET, TRIO, QUARTET, QUINTET, 
    SEXTET, SEPTET, OCTET, NONET, DECTET;
    
    public int numberOfMusicians() {
        return ordinal() + 1;
    }
}
```

- 위 코드는 동작은 하지만 유지보수하기가 끔찍한 코드
- 상수 선언 순서를 바꾸는 순간 numberOfMusicians 메서드가 오동작
- 이미 사용 중인 정수와 값이 같은 상수는 추가할 방법이 없음.
  - ex) 8중주 (octeet) 상수가 이미 있으니 똑같이 8명이 연주하는 복4중주(double quartet)는 추가할 수 없음.
- 값을 중간에 비워둘 수 없음
  - ex) 12명이 연주하는 3중 4중주 (triple quartet)을 추가하는 경우 중간에 11명짜리 상수도 채워야 하는데 그러한 연주를 일컫는 이름이 존재하지 않음
  - 이러한 문제를 해결하기 위해 쓰이지 않는 dummy 상수를 추가해야하는데 코드가 더러워지며 실용성이 떨어짐

### 해결책
- 열거 타입 상수에 연결된 값은 ordinal 메서드로 얻지 말고 인스턴스 필드에 저장

## ordinal 메서드 -> 인스턴스 필드에 저장

``` java
/**
 * @author : 조재철
 * @since 1.0
 */
public enum Ensemble {
    SOLO(1), DUET(2), TRIO(3), QUARTET(4), QUINTET(5),
    SEXTET(6), SEPTET(7), OCTET(8), DOUBLE_QUARTET(8),
    NONET(9), DECTET(1), TRIPLE_QUARTET(12);
    
    private final int numberOfMusicians;

    Ensemble(int size) {
        this.numberOfMusicians = size;
    }
    
    public int numberOfMusicians() {
        return this.numberOfMusicians;
    }
}

```

### 열거 타입 상수에 연결된 값을 인스턴스 필드에 저장시 주의 사항
<img width="618" alt="image" src="https://user-images.githubusercontent.com/90208100/215308832-236ca857-85ce-434f-824c-a7404b0e317a.png">
<img width="396" alt="image" src="https://user-images.githubusercontent.com/90208100/215308843-7b224d75-a7a9-4683-ba20-0af83c84f917.png">

- 그냥 인스턴스 필드에 저장만 한다면 위와 같은 에러가 일어남
- 열거 타입 상수와 연결된 정수값을 나타낼 필드를 두고 해당 필드를 초기화 할 생성자를 만들어야 함!!

### Enum의 API 문서에서 ordinal 메서드 설명
- ```대부분 프로그래머는 이 메서드를 사용할 일이 없다``` -> 이 메서드는 EnumSet과 EnumMap 같이 열거 타입 기반의 범용 자료구조에 쓸 목적으로 설계 되었던 것임
- 따라서 위의 용도가 아니라면 ordinal 메서드를 절대 사용하지 말자!!!
- EnumSet, EnumMap경우 아이템36과 아이템 37을 참고하자 

<img width="1176" alt="image" src="https://user-images.githubusercontent.com/90208100/215309449-735bd471-24cc-4d0a-b67c-3d919c638fcc.png">

- [참고 java11 doc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Enum.html)

**Comment by Irisation23** [comment, marked as answer]
정리해주신 내용 잘 봤습니다 😁

혹시 필드를 하나만 두고있는 enum 클래스의 경우 다르게 관리하고 있는 내용은 없을까에 대해 궁금해서 찾아보니
필드가 두개인 점은 다르지만, 배민에 계셨던 이동욱씨의 정리가 나오더군요. 

<img width="675" alt="image" src="https://user-images.githubusercontent.com/53285909/215347640-0a55ff85-a6cf-426f-8486-326bbd5971ed.png">

- 내용 발췌 [Java Enum 활용기 | 우아한형제들](https://techblog.woowahan.com/2527/)

@JoisFe 님이 사용하셨던 `numberOfMusicians()` 해당 메서드는 lombok의 `@Getter`로  대체해도 좋은 코드가 될 것 같다는 생각이 드네요!

- lombok is GOAT 🐐

**Reply by JoisFe** [reply]
정리글의 단순히 특정 정수값 하나 뿐만 아니라 여러 값을 묶을 수도 있는 예시 감사합니다.
